### ISSUE TITLE
0.5.4 crashing with PANIC "Milestones missing between (2272659) and (2272661)"

### ISSUE THREAD

**Issue by c0deright** [iotaledger/hornet#739]
**Describe the bug**

Did run 0.5.3 up until it did hang at 2272659. Did shut it down and started 0.5.4.

Using the docker-image

**Environment information:**
 - HORNET version 0.5.4 via docker

**Additional context**
```
hornet_1  |               ██╗  ██╗ ██████╗ ██████╗ ███╗   ██╗███████╗████████╗
hornet_1  |               ██║  ██║██╔═══██╗██╔══██╗████╗  ██║██╔════╝╚══██╔══╝
hornet_1  |               ███████║██║   ██║██████╔╝██╔██╗ ██║█████╗     ██║
hornet_1  |               ██╔══██║██║   ██║██╔══██╗██║╚██╗██║██╔══╝     ██║
hornet_1  |               ██║  ██║╚██████╔╝██║  ██║██║ ╚████║███████╗   ██║
hornet_1  |               ╚═╝  ╚═╝ ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═══╝╚══════╝   ╚═╝
hornet_1  |                                    v0.5.4
hornet_1  |
hornet_1  |
hornet_1  | 2020-11-13T12:52:55Z        INFO    CLI     Profile mode 'auto', Using profile '8gb'
hornet_1  | 2020-11-13T12:52:55Z        INFO    CLI     Loading plugins ...
hornet_1  | 2020-11-13T12:52:55Z        INFO    Node    Loading Plugin: CLI ... done
hornet_1  | 2020-11-13T12:52:55Z        INFO    Node    Loading Plugin: Graceful Shutdown ... done
hornet_1  | 2020-11-13T12:52:55Z        INFO    Node    Loading Plugin: Profiling ... done
hornet_1  | 2020-11-13T12:52:55Z        WARN    Database        Deleting invalid milestone 2272660 (JVYIDMDDFISYRAMQL9W9YQTDSWKVOLHOISWTFDVNNQGZAQUYLWMYYMLXQQCVVVXJWHWG9ULP9QEJJUQDC)
hornet_1  | 2020-11-13T12:52:55Z        WARN    Database        Deleting invalid milestone 2272661 (RKAZMTNNOOZGEIS9SYUVLFRRRRFWALTHSSKZPVNQPSCIX9CYHTIJXNWBXVQECJYLJKY99YMWIBBJXILXC)
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: Database ... done
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: Curl ... done
hornet_1  | 2020-11-13T12:52:56Z        INFO    Local   Initialized local: peer://DxPMDJRQrG4LCQ95gH7fC98NU3d95p5u9YvtKpRW1wd9@0.0.0.0:14626
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: Autopeering ... done
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: WebAPI ... done
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: PoW ... done
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: Gossip ... done
hornet_1  | SnapshotInfo:
hornet_1  |     CooAddr: UDYXTZBE9GZGPM9SSQV9LTZNDLJIZMPUVVXYXFYVBLIEUHLSEWFTKZZLXYRHHWVQV9MNNX9KZC9D9UZWZ
hornet_1  |     SnapshotIndex: 2272169 (YKIDVOOWSHWYTKPUMOEGOMLCUPGIMIQGLRJFYNBIDLGSNVZOUHOAODYQVWIZKNURWUXCJGBEGSTW99999)
hornet_1  |     EntryPointIndex: 2212085
hornet_1  |     PruningIndex: 2212085
hornet_1  |     Timestamp: 2020-11-12 18:12:28 +0000 UTC
hornet_1  |     SpentAddressesEnabled: true
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: Tangle ... done
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: Peering ... done
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: WarpSync ... done
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: URTS ... done
hornet_1  | 2020-11-13T12:52:56Z        INFO    Node    Loading Plugin: Metrics ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Loading Plugin: Snapshot ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Loading Plugin: Dashboard ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Skipping Plugin: ZMQ
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Skipping Plugin: MQTT
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Skipping Plugin: Spammer
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Skipping Plugin: Coordinator
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Skipping Plugin: Prometheus
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Executing plugins ...
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Starting Plugin: CLI ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Starting Plugin: Graceful Shutdown ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Starting Plugin: Profiling ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Starting Plugin: Database ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Starting Plugin: Curl ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Starting Plugin: Autopeering ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    WebAPI  Starting WebAPI server ...
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Starting Plugin: WebAPI ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Starting Plugin: PoW ... done
hornet_1  | 2020-11-13T12:53:05Z        INFO    Node    Starting Plugin: Gossip ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Starting TangleProcessor ...
hornet_1  | 2020-11-13T12:53:07Z        INFO    Node    Starting Plugin: Tangle ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Node    Starting Plugin: Peering ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Node    Starting Plugin: WarpSync ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Node    Starting Plugin: URTS ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Node    Starting Plugin: Metrics ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Node    Starting Plugin: Snapshot ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Dashboard       You can now access the dashboard using: http://localhost:8081
hornet_1  | ⇨ http server started on 127.0.0.1:8081
hornet_1  | 2020-11-13T12:53:07Z        INFO    Node    Starting Plugin: Dashboard ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Node    Starting background workers ...
hornet_1  | 2020-11-13T12:53:07Z        INFO    WebAPI  Starting WebAPI server ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    WebAPI  You can now access the API using: http://0.0.0.0:14265
hornet_1  | 2020-11-13T12:53:07Z        INFO    Gossip  Running BroadcastQueue
hornet_1  | 2020-11-13T12:53:07Z        INFO    Gossip  Running MessageProcessor
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Starting TangleProcessor[ProcessMilestone] ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    PoW     Starting PoW Handler ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Curl    Starting Curl batched hashing ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Peering Peering Server (0.0.0.0:15600) ...
hornet_1  | 2020-11-13T12:53:07Z        INFO    Peering Peering Server (0.0.0.0:15600) ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Autopeering
hornet_1  |
hornet_1  | WARNING: The autopeering plugin will disclose your public IP address to possibly all nodes and entry points. Please disable this plugin if you do not want this to happen!
hornet_1  |
hornet_1  | 2020-11-13T12:53:07Z        INFO    Autopeering     salt updated; expires=13 Nov 20 13:23 UTC
hornet_1  | 2020-11-13T12:53:07Z        INFO    Autopeering     started: ID=KDYjfdUmxrK Address=0.0.0.0:14626/udp PublicKey=DxPMDJRQrG4LCQ95gH7fC98NU3d95p5u9YvtKpRW1wd9
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Starting TangleProcessor[ReceiveTx] ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Snapshot        Starting LocalSnapshots ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Starting TangleProcessor[MilestoneSolidifier] ... done
hornet_1  | 2020-11-13T12:53:07Z        INFO    Peering trying to connect to 3 peers
hornet_1  | 2020-11-13T12:53:07Z        INFO    Autopeering     discovered: 88.99.87.135:14626 / eWj93Ci2BLr
hornet_1  | 2020-11-13T12:53:07Z        INFO    Peering handshaking with 207.180.242.11:15600...
hornet_1  | 2020-11-13T12:53:07Z        INFO    Peering handshaking with 35.171.233.10:15600...
hornet_1  | 2020-11-13T12:53:07Z        WARN    Peering error read tcp 172.19.0.2:55914->207.180.242.11:15600: read: connection reset by peer
hornet_1  | 2020-11-13T12:53:07Z        INFO    Peering moved disconnected 207.180.242.11:15600 into the reconnect pool
hornet_1  | 2020-11-13T12:53:07Z        INFO    Peering handshaking with 213.136.79.161:15600...
hornet_1  | 2020-11-13T12:53:07Z        INFO    Peering connected with 213.136.79.161:15600 [alias: dterango#6389] [feature set(s): Chrysalis-Pt1]
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Valid milestone detected! Index: 2273189, Hash: FUQJOHTIWOFHERDWTZBDVOLRWBGUTJAXROMAYEYEALLPMHZAIMAEOOSWGBDKGWWZPHASQZONVNHXA9999
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Run solidity check for Milestone (2272662)...
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Valid milestone detected! Index: 2272661, Hash: ZZPEAOXFDIZHOWXDWYNTYWIYKSPMAWJTVJYJWTR9VIHNAFOTMVXSCGLRJTDCNXWBXNSFNLAFIATCZ9999
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Solidifier finished: txs: 8, collect: 0s, solidity 3ms, propagation: 0s, total: 3ms
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Run solidity check for Milestone (2272661)...
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Solidifier finished: txs: 0, collect: 0s, solidity 0s, propagation: 0s, total: 0s
hornet_1  | 2020-11-13T12:53:07Z        PANIC   Tangle  Milestones missing between (2272659) and (2272661).
hornet_1  | github.com/gohornet/hornet/plugins/tangle.solidifyMilestone
hornet_1  |     /__w/hornet/hornet/plugins/tangle/solidifier.go:413
hornet_1  | github.com/gohornet/hornet/plugins/tangle.configureTangleProcessor.func3
hornet_1  |     /__w/hornet/hornet/plugins/tangle/tangle_processor.go:49
hornet_1  | github.com/iotaledger/hive.go/workerpool.(*WorkerPool).startWorkers.func1
hornet_1  |     /github/home/go/pkg/mod/github.com/iotaledger/hive.go@v0.0.0-20201016154508-2514b782563a/workerpool/workerpool.go:158
hornet_1  | 2020-11-13T12:53:07Z        INFO    Tangle  Run solidity check for Milestone (2272661)...
hornet_1  | panic: Milestones missing between (2272659) and (2272661).
hornet_1  |
hornet_1  | goroutine 217 [running]:
hornet_1  | go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000132d10, 0x0, 0x0, 0x0)
hornet_1  |     /github/home/go/pkg/mod/go.uber.org/zap@v1.15.0/zapcore/entry.go:230 +0x55f
hornet_1  | go.uber.org/zap.(*SugaredLogger).log(0xc0009852c8, 0xc004e5d104, 0x133fb16, 0x29, 0xc004071df8, 0x2, 0x2, 0x0, 0x0, 0x0)
hornet_1  |     /github/home/go/pkg/mod/go.uber.org/zap@v1.15.0/sugar.go:234 +0xf6
hornet_1  | go.uber.org/zap.(*SugaredLogger).Panicf(...)
hornet_1  |     /github/home/go/pkg/mod/go.uber.org/zap@v1.15.0/sugar.go:159
hornet_1  | github.com/gohornet/hornet/plugins/tangle.solidifyMilestone(0x100000000)
hornet_1  |     /__w/hornet/hornet/plugins/tangle/solidifier.go:413 +0x5f4
hornet_1  | github.com/gohornet/hornet/plugins/tangle.configureTangleProcessor.func3(0xc00039b700, 0x2, 0x2, 0xc000a071a0)
hornet_1  |     /__w/hornet/hornet/plugins/tangle/tangle_processor.go:49 +0x76
hornet_1  | github.com/iotaledger/hive.go/workerpool.(*WorkerPool).startWorkers.func1(0xc003769310)
hornet_1  |     /github/home/go/pkg/mod/github.com/iotaledger/hive.go@v0.0.0-20201016154508-2514b782563a/workerpool/workerpool.go:158 +0x1b6
hornet_1  | created by github.com/iotaledger/hive.go/workerpool.(*WorkerPool).startWorkers
hornet_1  |     /github/home/go/pkg/mod/github.com/iotaledger/hive.go@v0.0.0-20201016154508-2514b782563a/workerpool/workerpool.go:134 +0x5f
```

## After automatic restart of docker container

```
hornet_1  | 2020-11-13T12:53:09Z        WARN    Tangle  HORNET was not shut down correctly, the database may be corrupted. Starting revalidation...
hornet_1  | 2020-11-13T12:53:09Z        INFO    Tangle  reverting database state back from 2273187 to local snapshot 2272169 (this might take a while)...
hornet_1  | 2020-11-13T12:53:11Z        INFO    Tangle  deleting milestones...80/795 (10.06%). 17s left...
hornet_1  | 2020-11-13T12:53:13Z        INFO    Tangle  deleting milestones...195/795 (24.53%). 12s left...
hornet_1  | 2020-11-13T12:53:15Z        INFO    Tangle  deleting milestones...250/795 (31.45%). 13s left...
hornet_1  | 2020-11-13T12:53:17Z        INFO    Tangle  deleting milestones...374/795 (47.04%). 9s left...
hornet_1  | 2020-11-13T12:53:19Z        INFO    Tangle  deleting milestones...490/795 (61.64%). 6s left...
hornet_1  | 2020-11-13T12:53:21Z        INFO    Tangle  deleting milestones...606/795 (76.23%). 3s left...
hornet_1  | 2020-11-13T12:53:23Z        INFO    Tangle  deleting milestones...713/795 (89.69%). 1s left...
hornet_1  | 2020-11-13T12:53:25Z        INFO    Tangle  deleting milestones...795/795 (100.00%) done. took 16.099s
hornet_1  | 2020-11-13T12:53:27Z        INFO    Tangle  analyzed 86942 ledger diffs
hornet_1  | 2020-11-13T12:53:28Z        INFO    Tangle  deleting ledger diffs...0/0 (100.00%) done. took 3.595s
hornet_1  | 2020-11-13T12:53:30Z        INFO    Tangle  analyzed 7461 transactions
hornet_1  | 2020-11-13T12:53:32Z        INFO    Tangle  analyzed 14953 transactions
hornet_1  | 2020-11-13T12:53:34Z        INFO    Tangle  analyzed 22746 transactions
hornet_1  | 2020-11-13T12:53:36Z        INFO    Tangle  analyzed 30635 transactions
hornet_1  | 2020-11-13T12:53:38Z        INFO    Tangle  analyzed 38276 transactions
hornet_1  | 2020-11-13T12:53:40Z        INFO    Tangle  analyzed 41023 transactions
hornet_1  | 2020-11-13T12:53:42Z        INFO    Tangle  analyzed 45625 transactions
hornet_1  | 2020-11-13T12:53:44Z        INFO    Tangle  analyzed 53162 transactions
hornet_1  | 2020-11-13T12:53:46Z        INFO    Tangle  analyzed 60620 transactions
[..]
```

**Comment by c0deright** [comment]
I removed peer.db, snapshot.db, spent.db and tangle.db so that hornet did a restore from the latest snapshot from before the COO crashed. Didn't see a hornet crash again and the node is synced and healthy again.

**Comment by EasonC13** [comment]
I enconter this problem when I upgrade from 0.5.3 to 0.5.5
So I change the version back to 0.5.4 and after revalidating the database, the node is sync now.

I think there is a better solution, though.

**Comment by muXxer** [comment]
Thx for the bug report @c0deright .
We fixed this with v0.5.5.

I wasn't able to test the fix myself, since I don't have an affected database, but I saw this fix working already for several other people with version 0.5.5.
Are you really sure that you were running the 0.5.5 after you upgraded @EasonC13 ?
Can you please provide full logs of the crash?

**Comment by muXxer** [comment]
Fixed with #744
